pallets/phala-world/src/nft_sale.rs

```diff
@@ -0,0 +1,1686 @@
+	///
+	/// Parameters:
+	/// - `status`: Status to set CanClaimSpirits StorageValue
+	fn set_claim_spirits_status(status: bool) -> DispatchResult {
+		<CanClaimSpirits<T>>::put(status);
+
+		Self::deposit_event(Event::ClaimSpiritStatusChanged { status });
+
+		Ok(())
+	}
+
+	/// Set Rare Origin of Shells status for purchase with the Overlord Admin Account to allow
+	/// users to purchase either Legendary or Magic Origin of Shells
+	///
+	/// Parameters:
+	/// `status`: Status to set CanPurchaseRareOriginOfShells StorageValue
+	fn set_purchase_rare_origin_of_shells_status(status: bool) -> DispatchResult {
+		<CanPurchaseRareOriginOfShells<T>>::put(status);
+
+		Self::deposit_event(Event::PurchaseRareOriginOfShellsStatusChanged { status });
+
+		Ok(())
+	}
+
+	/// Set Hero Origin of Shells status for purchase with the Overlord Admin Account to allow
+	/// users to purchase Hero Origin of Shells
+	///
+	/// Parameters:
+	/// `status`: Status to set CanPurchaseOriginOfShellsWhitelist StorageValue
+	fn set_purchase_hero_origin_of_shells_status(status: bool) -> DispatchResult {
+		<CanPurchaseHeroOriginOfShells<T>>::put(status);
+
+		Self::deposit_event(Event::PurchaseHeroOriginOfShellsStatusChanged { status });
+
+		Ok(())
+	}
+
+	/// Set status of Preordering origin_of_shells with the Overlord Admin Account to allow
+	/// users to preorder origin_of_shells through the `preorder_origin_of_shell()` function
+	///
+	/// Parameters:
+	/// - `status`: Status to set CanPreorderOriginOfShells StorageValue
+	fn set_preorder_origin_of_shells_status(status: bool) -> DispatchResult {
+		<CanPreorderOriginOfShells<T>>::put(status);
+
+		Self::deposit_event(Event::PreorderOriginOfShellsStatusChanged { status });
+
+		Ok(())
+	}
+
+	/// Set status of last day of sale for origin of shells with the Overlord Admin Account to allow
+	/// users to purchase any origin of shell
+	///
+	/// Parameters:
+	/// - `status`: Status to set LastDayOfSale StorageValue
+	fn set_last_day_of_sale_status(status: bool) -> DispatchResult {
+		<LastDayOfSale<T>>::put(status);
+
+		Self::deposit_event(Event::PreorderOriginOfShellsStatusChanged { status });
+
+		Ok(())
+	}
+
+	/// Set initial OriginOfShellInventory values in the StorageDoubleMap. Key1 will be of
+	/// OriginOfShellType and Key2 will be the RaceType and the Value will be NftSaleInfo struct
+	/// containing the information for the NFT sale. Initial config will look as follows:
+	/// `<Legendary>,<RaceType> => NftSaleInfo { race_count: 0, career_count: 0,
+	/// race_for_sale_count: 1, race_giveaway_count: 0, race_reserved_count: 1 }`
+	/// `<Magic>,<RaceType> => NftSaleInfo { race_count: 0, career_count: 0, race_for_sale_count:
+	/// 15, race_giveaway_count: 0, race_reserved_count: 5 }`
+	/// `<Hero>,<RaceType> => NftSaleInfo { race_count: 0, career_count: 0, race_for_sale_count:
+	/// 1250, race_giveaway_count: 50, race_reserved_count: 0 }`
+	fn set_initial_origin_of_shell_inventory() -> DispatchResult {
```

Comment:
Is it completed yet?

Reply:
This gets executed by `Overlord` account in the `init.js` script. This sets the numbers for the NFT Sale up until the whitelist purchases are finished. After that, we calculate the remaining numbers based on the claimed `Hero` Origin of Shells and call `update_origin_of_shell_type_counts` with `Overlord` account to set the remaining NFTs available for the Preorders.

Reply:
These numbers are based on the update for how much we allocate for each Origin of Shell Type and Race. Are these numbers still valid?

Reply:
Got it. I think it's still valid. Would be nice to double check with the team